### PR TITLE
More optimization on TopK functions.

### PR DESCRIPTION
**Pull request by big-andy-coates**
Fix for #880.

PR looks to:
- early out of `aggregate` even earlier.
- remove unnecessary array creation `Merger`
- reduce size of required array in `Merger`
- zip the two source arrays together, taking advantage of the fact the source topK arrays are already sorted.
- remove unused code from `ArrayUtil`

Note: This change does rely on the fact that the source arrays passing into the `Merger` are already sorted.  I assume that the `Merger` will always be passed source arrays created by itself, and so this is a safe assumption.

### CODE REVIEW

File: ksql-engine/src/test/java/io/confluent/ksql/function/udaf/topkdistinct/IntTopkDistinctKudafTest.java [hunk 1]
```diff
@@ -134,4 +134,48 @@ public void shouldBeThreadSafe() {
...
+    final int topX = 10;
+    final TopkDistinctKudaf<Integer> intTopkDistinctKudaf =
+        new TopkDistinctKudaf<>(0, topX, Integer.class);
+    final Integer[] aggregate = IntStream.range(0, topX)
+        .mapToObj(idx -> null)
+        .toArray(Integer[]::new);
+    final long start = System.currentTimeMillis();
+
+    for(int i = 0; i != iterations; ++i) {
+      intTopkDistinctKudaf.aggregate(i, aggregate);
```
Comment (rodesai) [inline ksql-engine/src/test/java/io/confluent/ksql/function/udaf/topkdistinct/IntTopkDistinctKudafTest.java]: Does this test take advantage of your optimization? The check for whether we can skip the sort will always be false right?

File: ksql-engine/src/test/java/io/confluent/ksql/function/udaf/topk/StringTopkKudafTest.java [hunk 2]
```diff
@@ -66,8 +67,8 @@ public void shouldAggregateTopKWithLessThanKValues() {
   public void shouldMergeTopK() {
     KsqlAggregateFunction<Object, Object[]> topkKudaf =
             topKFactory.getProperAggregateFunction(argumentType);
-    String[] array1 = new String[]{"123", "Hello", "paper"};
-    String[] array2 = new String[]{"Hi", "456", "Zzz"};
+    String[] array1 = new String[]{"paper", "Hello", "123"};
+    String[] array2 = new String[]{"Zzz", "Hi", "456"};
 
     assertThat("Invalid results.", topkKudaf.getMerger().apply("key", array1, array2),
```
Comment (rodesai) [inline ksql-engine/src/test/java/io/confluent/ksql/function/udaf/topk/StringTopkKudafTest.java]: Not really in scope of your PR but this is kind of weird. I think other systems order strings alphabetically.

Comment (big-andy-coates) [reply ksql-engine/src/test/java/io/confluent/ksql/function/udaf/topk/StringTopkKudafTest.java]: Yer, our's uses reverse order. TopX on strings is a bit weird anyway.  Maybe we need `BottomX` too!

File: ksql-engine/src/test/java/io/confluent/ksql/function/udaf/topkdistinct/IntTopkDistinctKudafTest.java (hunk 1 above)
Comment (big-andy-coates) [reply ksql-engine/src/test/java/io/confluent/ksql/function/udaf/topkdistinct/IntTopkDistinctKudafTest.java]: Nope, I was testing worst case to make sure I was improving things!

**Comment by big-andy-coates** [comment]
@apurvam yep, totally agree.

**Pull request merged**